File: bamboo/src/main/java/capsule/bamboo/Logger.java
package capsule.bamboo;

import android.util.Log;

/**
 * Created by hhly-pc on 2016/12/23.
 */

public class Logger {

  private static final String DEFAULT_TAG = "vegeta";

  private static String  tag       = DEFAULT_TAG;
  private static boolean logEnable = true;

  public static void configTag(String tag) {
    Logger.tag = tag;
  }

  public static void configLogEnable(boolean logEnable) {
    Logger.logEnable = logEnable;
  }

  /* ******* verbose ****** */

  public static void v(String msg) {
    v(null, msg);
  }

  public static void v(Object clazz, String msg) {
    Class c = clazz.getClass();
    String className = c.getName();
    className = className.substring(className.lastIndexOf("."));
    v(className, msg);
  }

  public static void v(String tag, String msg) {
    String t = checkTag(tag);
    printLog("v", t, msg);
  }
  /* ******* debug ****** */

  public static void d(String msg) {
    d(null, msg);
  }

  public static void d(Object clazz, String msg) {
    Class c = clazz.getClass();
    String className = c.getName();
    className = className.substring(className.lastIndexOf("."));
    d(className, msg);
  }

  public static void d(String tag, String msg) {
    String t = checkTag(tag);
    printLog("d", t, msg);
  }
  /* ******* information ****** */

  public static void i(String msg) {
    i(null, msg);
  }

  public static void i(Object clazz, String msg) {
    Class c = clazz.getClass();
    String className = c.getName();
    className = className.substring(className.lastIndexOf("."));
    i(className, msg);
  }

  public static void i(String tag, String msg) {
    String t = checkTag(tag);
    printLog("i", t, msg);
  }
  /* ******* warning ****** */

  public static void w(String msg) {
    w(null, msg);
  }

  public static void w(Object clazz, String msg) {
    Class c = clazz.getClass();
    String className = c.getName();
    className = className.substring(className.lastIndexOf("."));
    w(className, msg);
  }

  public static void w(String tag, String msg) {
    String t = checkTag(tag);
    printLog("w", t, msg);
  }
  /* ******* error ****** */

  public static void e(String msg) {
    e(null, msg);
  }

  public static void e(Object clazz, String msg) {
    Class c = clazz.getClass();
    String className = c.getName();
    className = className.substring(className.lastIndexOf("."));
    e(className, msg);
  }

  public static void e(String tag, String msg) {
    String t = checkTag(tag);
    printLog("e", t, msg);
  }

  /* ****** internal method ****** */

  private static String checkTag(String tag) {
    return tag == null ? Logger.tag : tag;
  }

  private static void printLog(String level, String tag, String msg) {
    if (!logEnable) {
      return;
    }
    String clickableMessage = buildClickableMessage(level, msg);
    switch (level) {
      case "v":
        Log.v(tag, clickableMessage);
        break;
      case "d":
        Log.d(tag, clickableMessage);
        break;
      case "i":
        Log.i(tag, clickableMessage);
        break;
      case "w":
        Log.w(tag, clickableMessage);
        break;
      case "e":
        Log.e(tag, clickableMessage);
        break;
    }
  }

  private static String buildClickableMessage(String level, String msg) {
    StackTraceElement[] stackTraceElement = Thread.currentThread().getStackTrace();
    int currentIndex = -1;
    for (int i = 0; i < stackTraceElement.length; i++) {
      if (stackTraceElement[i].getMethodName().compareTo(level) == 0) {
        currentIndex = i;
      }
    }
    currentIndex++;

    String fullClassName = stackTraceElement[currentIndex].getClassName();
    String className = fullClassName.substring(fullClassName.lastIndexOf(".") + 1);
    String methodName = stackTraceElement[currentIndex].getMethodName();
    String lineNumber = String.valueOf(stackTraceElement[currentIndex].getLineNumber());

    return msg
        + "\nat "
        + fullClassName
        + "."
        + methodName
        + "("
        + className
        + ".java:"
        + lineNumber
        + ")";
  }
}

File: lib/src/main/java/yuluyao/logger/Logger.kt
package yuluyao.logger

import android.util.Log

/**
 * Created by hhly-pc on 2016/12/23.
 */
object Logger {

  //<editor-fold desc="config">
  private const val DEFAULT_TAG = "vegeta"

  var defaultTag = DEFAULT_TAG
  var logEnable = true
  //</editor-fold>



  //<editor-fold desc="verbose">
  fun v(msg: String) {
    v(null, msg)
  }

  fun v(tag: Any?, msg: String?) = when (tag) {
    null -> printLog(VERBOSE, defaultTag, msg)
    is String -> printLog(VERBOSE, tag, msg)
    else -> printLog(VERBOSE, tag.javaClass.simpleName, msg)
  }
  //</editor-fold>

  //<editor-fold desc="debug">
  fun d(msg: String) {
    d(null, msg)
  }

  fun d(tag: Any?, msg: String?) = when (tag) {
    null -> printLog(DEBUG, defaultTag, msg)
    is String -> printLog(DEBUG, tag, msg)
    else -> printLog(DEBUG, tag.javaClass.simpleName, msg)
  }
  //</editor-fold>

  //<editor-fold desc="info">
  fun i(msg: String) {
    i(null, msg)
  }

  fun i(tag: Any?, msg: String?) = when (tag) {
    null -> printLog(INFO, defaultTag, msg)
    is String -> printLog(INFO, tag, msg)
    else -> printLog(INFO, tag.javaClass.simpleName, msg)
  }
  //</editor-fold>

  //<editor-fold desc="warn">
  fun w(msg: String) {
    w(null, msg)
  }

  fun w(tag: Any?, msg: String?) = when (tag) {
    null -> printLog(WARN, defaultTag, msg)
    is String -> printLog(WARN, tag, msg)
    else -> printLog(WARN, tag.javaClass.simpleName, msg)
  }
  //</editor-fold>

  //<editor-fold desc="error">
  fun e(msg: String) {
    e(null, msg)
  }

  fun e(tag: Any?, msg: String?) = when (tag) {
    null -> printLog(ERROR, defaultTag, msg)
    is String -> printLog(ERROR, tag, msg)
    else -> printLog(ERROR, tag.javaClass.simpleName, msg)
  }
  //</editor-fold>

  //<editor-fold desc="internal">

  private const val VERBOSE = "v"
  private const val DEBUG = "d"
  private const val INFO = "i"
  private const val WARN = "w"
  private const val ERROR = "e"

  private fun printLog(level: String, tag: String, msg: String?) {
    if (!logEnable) {
      return
    }
    val clickableMessage = buildClickableMessage(level, msg)
    when (level) {
      VERBOSE -> Log.v(tag, clickableMessage)
      DEBUG -> Log.d(tag, clickableMessage)
      INFO -> Log.i(tag, clickableMessage)
      WARN -> Log.w(tag, clickableMessage)
      ERROR -> Log.e(tag, clickableMessage)
    }
  }

  private fun buildClickableMessage(level: String, msg: String?): String {
    val stackTraceElement = Thread.currentThread().stackTrace
    var currentIndex = -1
    for (i in stackTraceElement.indices) {
      if (stackTraceElement[i].methodName.compareTo(level) == 0) {
        currentIndex = i
      }
    }
    currentIndex++
    val fullClassName = stackTraceElement[currentIndex].className
    val className = fullClassName.substring(fullClassName.lastIndexOf(".") + 1)
    val methodName = stackTraceElement[currentIndex].methodName
    val lineNumber = stackTraceElement[currentIndex].lineNumber.toString()
    return """
      $msg
      at $fullClassName.$methodName($className.java:$lineNumber)
      """.trimIndent()
  }
  //</editor-fold>


}